src/support/qsort.c

```diff
@@ -0,0 +1,269 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "wt_internal.h"
+
+#define WT_QSORT_INSERT_THRESHOLD 7
+#define WT_QSORT_LARGE_THRESHOLD 40
+
+#ifdef WT_QSORT_R_DEF
+typedef int wt_cmp_t(const void *, const void *, void *);
+#else
+typedef int wt_cmp_t(const void *, const void *);
+#endif
+
+#ifdef WT_QSORT_R_DEF
+#define CMP(x, y, t) (cmp((x), (y), (t)))
+#else
+#define CMP(x, y, t) (cmp((x), (y)))
+#endif
+
+/*
+ * __swap_bytes --
+ *     Swap the contents of two arbitrary values of a given size.
+ */
+static inline void
+__swap_bytes(uint8_t *a, uint8_t *b, size_t size)
+{
+    uint8_t tmp;
+
+    while (size > 0) {
+        tmp = *a;
+        *a++ = *b;
+        *b++ = tmp;
+        --size;
+    }
+}
+
+/*
+ * __med3 --
+ *     Find the median of three values, using the user-supplied comparator.
+ */
+static inline void *
+__med3(void *a, void *b, void *c, wt_cmp_t cmp, void *context)
+{
+#ifndef WT_QSORT_R_DEF
+    WT_UNUSED(context);
+#endif
+
+    if (CMP(a, b, context) < 0) {
+        if (CMP(b, c, context) < 0)
+            return (b);
+        else
+            return (CMP(a, c, context) < 0 ? c : a);
+    } else {
+        if (CMP(b, c, context) > 0)
+            return (b);
+        else
+            return (CMP(a, c, context) < 0 ? a : c);
+    }
+}
+
+/*
+ * __qsort --
+ *     Sort an array using a user-supplied comparator function and a context argument.
+ */
+static void
+__qsort(void *arr, size_t nmemb, size_t elem_sz, wt_cmp_t cmp, void *context)
+{
+    /*
+     * This is originally from "Engineering a Sort Function", by Jon L. Bentley and Douglas
+     * M. McIlroy ("Software - Practice and Experience, Vol. 23(11)", November 1993). It's been
+     * massaged into a form suitable for WiredTiger, initially by Luke Pearson.
```

Comment:
This code looks very much like it was adapted from the FreeBSD project https://github.com/freebsd/freebsd-src/blob/main/lib/libc/stdlib/qsort.c as is evidenced by the copyright. This does not read that way imho.

Reply:
I'll have to fob that one off to @luke-pearson

Reply:
@marcbutler when you say "this" doesn't read that way, you mean this comment right? Its similar to the same one at the top of FreeBSD's qsort.c. We could also add a bit to this comment that it was adapted from FreeBSD's implementation? I think its fine to say that we adapted the code further for WiredTiger.

Reply:
@luke-pearson Yes the comment. The way this reads to me is that you adapted from paper (which I infer had example implementation). 

My understanding is that this code is taken from FreeBSD and adapted to this projects conventions. The FreeBSD implementation is taken from the paper and the reference is far more terse:
```
/*
 * Qsort routine from Bentley & McIlroy's "Engineering a Sort Function".
 */
```

Also I don't see other developers names embedded in the source. I would have thought it falls under a "co-authored" entry in the commit.

Reply:
Yes happy to have my name dropped from the comment and listed as 'coauthered'. I think what Will has done with the comment is otherwise good, and should also reference FreeBSD.

Reply:
Done in bdd1457, as a starting point - I'll try remember to put you as a co-author @luke-pearson.